pylint/checkers/refactoring/refactoring_checker.py

```diff
@@ -1601,3 +1614,167 @@ def _check_return_at_the_end(self, node):
+            inferred = utils.safe_infer(node.iter.func)
+            if not isinstance(inferred, astroid.BoundMethod):
+                return
+            iterating_object_name = node.iter.as_string().rpartition(".items")[0]
+
+            # Verify that the body of the for loop uses a subscript
+            # with the object that was iterated. This uses some heuristics
+            # in order to make sure that the same object is used in the
+            # for body.
+            for child in node.body:
+                for subscript in child.nodes_of_class(astroid.Subscript):
+                    subscript = cast(astroid.Subscript, subscript)
+
+                    if not isinstance(
+                        subscript.value, (astroid.Name, astroid.Attribute)
+                    ):
+                        continue
+
+                    value = subscript.slice
+                    if isinstance(value, astroid.Index):
+                        value = value.value
+
+                    if (
+                        isinstance(subscript.parent, astroid.Assign)
+                        and subscript in subscript.parent.targets
+                        or isinstance(subscript.parent, astroid.AugAssign)
+                        and subscript == subscript.parent.target
+                    ):
+                        # Ignore this subscript if it is the target of an assignment
+                        continue
+
+                    # Case where .items is assigned to k,v (i.e., for k, v in d.items())
+                    if isinstance(value, astroid.Name):
+                        if (
+                            not isinstance(node.target, astroid.Tuple)
+                            or value.name != node.target.elts[0].name
+                            or iterating_object_name != subscript.value.as_string()
+                        ):
+                            continue
+
+                        last_definition_lineno = value.lookup(value.name)[1][-1].lineno
+                        if last_definition_lineno > node.lineno:
+                            # Ignore this subscript if it has been redefined after
+                            # the for loop. This checks for the line number using .lookup()
+                            # to get the line number where the iterating object was last
+                            # defined and compare that to the for loop's line number
+                            continue
+
+                        self.add_message(
+                            "unnecessary-dict-indexing",
+                            node=subscript,
+                            args=(node.target.elts[1].as_string()),
+                        )
+
+                    # Case where .items is assigned to single var (i.e., for item in d.items())
+                    elif isinstance(value, astroid.Subscript):
+                        # Note: We don't have to check for reassignment for this case as
+                        # tuples are immutable
```

Comment:
Have you tried it? The tuple itself might be immutable, but the name can be reassigned without issues.
```py
for item in d:
  d[item[0]]  # add warning
  item = (1, "a")
  d[item[0]]  # fine
```

Reply:
Ah okay, thanks for pointing that out. I did this and missed it
```python
for item in d:
  d[item[0]]  # add warning
  item[0] = 3
  d[item[0]]  # fine
```

Reply:
resolved in 14c1c2fd